ssl/quic/quic_port.c

```diff
@@ -620,6 +647,154 @@ static int port_try_handle_stateless_reset(QUIC_PORT *port, const QUIC_URXE *e)
+ *
+ * This function performs the following steps:
+ * - Generates a validation token for the client.
+ * - Sets the destination and source connection IDs.
+ * - Calculates the integrity tag and sets the token length.
+ * - Encodes and sends the packet via the BIO network interface.
+ *
+ * Error handling is included for failures in CID generation, encoding, and
+ * network transmiss
+ */
+static void port_send_retry(QUIC_PORT *port,
+                            BIO_ADDR *peer,
+                            QUIC_PKT_HDR *client_hdr)
+{
+    BIO_MSG msg[1];
+    unsigned char buffer[512];
+    WPACKET wpkt;
+    size_t written;
+    QUIC_PKT_HDR hdr;
+    struct validation_token token;
+    size_t token_len = TOKEN_LEN;
+    unsigned char *integrity_tag;
+    int ok;
+
+    /* TODO(QUIC_SERVER): generate proper validation token */
+    memcpy(token.token_buf, "openssltoken", sizeof("openssltoken") - 1);
+
+    token.token_odcid = client_hdr->dst_conn_id;
+    token_len += token.token_odcid.id_len;
+    integrity_tag = (unsigned char *)&token.token_odcid +
+        token.token_odcid.id_len + sizeof(token.token_odcid.id_len);
+    /*
+     * 17.2.5.1 Sending a Retry packet
+     *   dst ConnId is src ConnId we got from client
+     *   src ConnId comes from local conn ID manager
+     */
+    memset(&hdr, 0, sizeof(QUIC_PKT_HDR));
+    hdr.dst_conn_id = client_hdr->src_conn_id;
+    /*
+     * this is the random connection ID, we expect client is
+     * going to send the ID with next INITIAL packet which
+     * will also come with token we generate here.
+     */
+    ok = ossl_quic_lcidm_get_unused_cid(port->lcidm, &hdr.src_conn_id);
+    if (ok == 0)
+        return;
+
+    hdr.dst_conn_id = client_hdr->src_conn_id;
+    hdr.type = QUIC_PKT_TYPE_RETRY;
+    hdr.fixed = 1;
+    hdr.version = 1;
+    hdr.len = token_len;
+    hdr.data = (unsigned char *)&token;
+    ok = ossl_quic_calculate_retry_integrity_tag(port->engine->libctx,
+                                                 port->engine->propq, &hdr,
+                                                 &client_hdr->dst_conn_id,
+                                                 integrity_tag);
+    if (ok == 0)
+        return;
+
+    hdr.token = (unsigned char *)&token;
+    hdr.token_len = token_len;
+
+    msg[0].data = buffer;
+    msg[0].peer = peer;
+    msg[0].local = NULL;
+    msg[0].flags = 0;
+
+    ok = WPACKET_init_static_len(&wpkt, buffer, sizeof(buffer), 0);
+    if (ok == 0)
+        return;
+
+    ok = ossl_quic_wire_encode_pkt_hdr(&wpkt, client_hdr->dst_conn_id.id_len,
+                                       &hdr, NULL);
+    if (ok == 0)
+        return;
+
+    ok = WPACKET_get_total_written(&wpkt, &msg[0].data_len);
+    if (ok == 0)
+        return;
+
+    ok = WPACKET_finish(&wpkt);
+    if (ok == 0)
+        return;
+
+    if (!BIO_sendmmsg(port->net_wbio, msg, sizeof(BIO_MSG), 1, 0, &written))
+        ERR_raise_data(ERR_LIB_SSL, SSL_R_QUIC_NETWORK_ERROR,
```

Comment:
I've just noticed the same thing at PR which introduces version negotiation. I have a same question here: is it worth to raise error here? If needs to be handled then I would prefer to go with trace/log event.

Reply:
A trace event might be better, as the ERR api saves information to the stack of the calling thread, and thats non-deterministic here

Reply:
followup question, should we have a QUIC trace category here?  The closest category we have now is TLS, and that doesn't seem to fit the bill here

Reply:
Hmm... this is even more problematic. The bio is non-blocking and it might fail with EAGAIN expecting that you call the sendmmsg() again when it becomes writable. So there is definitely at least TODO(QUIC_SERVER) to resolve here.

Reply:
there is also a qlog subsystem for QUIC. however not sure if there is any mean to signal/log failure in sedmmsg()
https://quicwg.org/qlog/draft-ietf-quic-qlog-quic-events.html#name-version_information

Reply:
qlog would make sense, but we're not really well positioned to use it at the moment.  qlog seems to be built around the concept of a channel (i.e. QUIC_CHANNELS hold qlog objects and report based on the channel they are connected to).  We could I suppose add a qlog object to a QUIC_PORT, add the appropriate QLOG_EVENT for port events, and add it in here, but I would need to investigate the complexity of such a refactor

Reply:
based on the stand up today, I think we're in consensus that addressing this logging issue should be moved to its own ticket to be addressed, as its a significant amount of additional work to get qlog working at the port level.  Tracked here:
https://github.com/openssl/project/issues/926

Reply:
@t8m add a comment with a TODO (QUIC SERVER) for the retry condition